docs/design-docs/02-custom-root-reconciler-clusterrole.md

```diff
@@ -0,0 +1,111 @@
+# Custom `ClusterRole` assignments for `RootSync`s
+
+* Author(s): @tomasaschan
+* Approver: @sdowell, @janetkuo
+* Status: approved
+
+## Summary
+
+When creating a reconciler deployment for a `RootSync`, Config Sync currently [creates a
+`ClusterRoleBinding`] for the generated `ServiceAccount` granting it `cluster-admin`.
+
+This proposal introduces a mechanism to customize this behavior.
+
+[creates a `ClusterRoleBinding`]: https://github.com/GoogleContainerTools/kpt-config-sync/blob/199db6dbfa0b9eb1824925c8c687574de095a294/pkg/reconcilermanager/controllers/rootsync_controller.go#L789
+
+## Motivation
+
+If one wants to allow deploying resources from a source where one doesn't have full
+control of the contents of the upstream config source, and want to limit what another
+team or third party can accomplish in the cluster, one can imagine limiting what
+permissions are granted to the `ServiceAccount` running the reconciler.
+
+However, as Config Sync is currently granting `cluster-admin`, any custom role bindings
+are effectively ignored; the reconciler will, due to the binding added by Config Sync,
+have access to do _anything_ regardless.
+
+To follow the [principle of least privilege], one should ensure the reconciler only has
+access to deploy the expected resources.
+
+See also [#935].
+
+[principle of least privilege]: https://en.wikipedia.org/wiki/Principle_of_least_privilege
+[#935]: https://github.com/GoogleContainerTools/kpt-config-sync/issues/935
+
+## Design Overview
+
+By providing the name of a user-defined `ClusterRole` for a `RootSync`, a user can
+override which role Config Sync binds to. This configuration is exposed as a new field
+on `spec.overrides` for a `RootSync`:
+
+```yaml
+spec:
+  overrides:
+    clusterRole: my-custom-role
+```
+
+This role is defaulted to `cluster-admin` in order to stay backwards compatible.
+
+For the case where a single `ClusterRole` is not expressive enough to configure the
+permissions a user want, we also allow the special sentinel value `%none%`, which
```

Comment:
I don't know; I don't think so. I picked it for two reasons:
1. It is a value that can never be a valid name of a `ClusterRole` (because `%` is a disallowed character), so setting it to this can never end up binding to an unintended role created later.
2. I found it more descriptive than e.g. using a single invalid character or something else along those lines.
If you don't like it we can remove that part of this PR; although I do find it appealing to have _some_ way to completely turn off the binding...

Reply:
Another option would be to make the property pointer-valued, and differentiate between "not set" and "set to an empty string", but that seems bad for other reasons...

Reply:
Another option could be to make this a struct with a nested `name` and `enabled` field. This could create a clearer API for how to disable ClusterRoleBinding, although adds more API permutations (e.g. `name=my-cluster-role, enabled=false` probably doesn't make sense).

I don't feel too strongly about it and am okay with `%none%`, my main concern is whether this is unconventional

Reply:
I had the struct idea independently (sleeping on ideas is often effective! 😄), but I imagined it as
```
type ClusterRoleBindingOverrides struct {
  RoleName string
  Disabled   bool
}
```
This way the zero value of the struct is the current behavior (if an empty `RoleName` defaults it to `"cluster-admin"`); setting the name to something else does what this PR currently does for any value other than `%none%`, and setting `Disabled: true` disables the binding (and ignores `RoleName`, i.e. `Disabled` takes precedence).

I'll work on an update here address this (and other comments).